Kubernetes 1.25 support (#593)
longhorn/website: content/docs/1.4.0/deploy/important-notes/index.md

### Recurring Jobs
After the upgrade, the recurring job settings of volumes will be migrated to new recurring job resources, and the `RecurringJobs` field in the volume spec will be deprecated. [[doc](https://longhorn.io/docs/{{< current-version >}}/deploy/upgrade/#4-automatically-migrate-recurring-jobs)]

### Backup Migration Time
When upgrading from a Longhorn version >= 1.2.0 and <= v1.2.3 to v{{< current-version >}}, if your cluster has many backups, you may expect to have a long upgrade time (with 22000 backups, it could take a few hours). Helm upgrade may timeout and you may need to re-run the upgrade command or set a longer timeout. This is [a known issue](https://github.com/longhorn/longhorn/issues/3890).

### Longhorn uninstallation
To prevent Longhorn from being accidentally uninstalled (which leads to data lost),
we introduce a new setting, [deleting-confirmation-flag](../../references/settings/#deleting-confirmation-flag).
If this flag is **false**, the Longhorn uninstallation job will fail.
Set this flag to **true** to allow Longhorn uninstallation.
See more in the [uninstall](../uninstall) section.

> 
> - Longhorn pods require privileged access to manage nodes' storage. In Longhorn `v1.3.x` or older, Longhorn was shipping some Pod Security Policies by default, (e.g., [link](https://github.com/longhorn/longhorn/blob/4ba39a989b4b482d51fd4bc651f61f2b419428bd/chart/values.yaml#L260)).
> However, Pod Security Policy has been deprecated since Kubernetes v1.21 and removed since Kubernetes v1.25, [link](https://kubernetes.io/docs/concepts/security/pod-security-policy/).
> Therefore, we stopped shipping the Pod Security Policies by default.
> For Kubernetes < v1.25, if your cluster still enables Pod Security Policy admission controller, set the helm value `enablePSP` to `true` to install `longhorn-psp` PodSecurityPolicy resrouce which allows privileged Longhorn pods to start.
> 
> - As a replacement for Pod Security Policy, Kubernetes provides a new mechanism, [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).
> If you enable the Pod Security Admission controller and change the default behavior to block privileged pods,
> you must add the correct labels to the namespace where Longhorn pods run to allow Longhorn pods to start successfully
> (because Longhorn pods require privileged access to manage storage).
> For example, adding the following labels to the namespace that is running Longhorn pods:
>     ```
>     apiVersion: v1
>     kind: Namespace
>     metadata:
>       name: longhorn-system
>       labels:
>         pod-security.kubernetes.io/enforce: privileged
>         pod-security.kubernetes.io/enforce-version: latest
>         pod-security.kubernetes.io/audit: privileged
>         pod-security.kubernetes.io/audit-version: latest
>         pod-security.kubernetes.io/warn: privileged

Reviewer: ```suggestion

### Pod Security Policies Disabled & Pod Security Admission Introduction

- Longhorn pods require privileged access to manage nodes' storage. In Longhorn `v1.3.x` or older, Longhorn was shipping some Pod Security Policies by default, (e.g., [link](https://github.com/longhorn/longhorn/blob/4ba39a989b4b482d51fd4bc651f61f2b419428bd/chart/values.yaml#L260)).
However, Pod Security Policy has been deprecated since Kubernetes v1.21 and removed since Kubernetes v1.25, [link](https://kubernetes.io/docs/concepts/security/pod-security-policy/).
Therefore, we stopped shipping the Pod Security Policies by default.
For Kubernetes < v1.25, if your cluster still enables Pod Security Policy admission controller, set the helm value `enablePSP` to `true` to install `longhorn-psp` PodSecurityPolicy resrouce which allows privileged Longhorn pods to start.

- As a replacement for Pod Security Policy, Kubernetes provides a new mechanism, [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).
If you enable the Pod Security Admission controller and change the default behavior to block privileged pods,
you must add the correct labels to the namespace where Longhorn pods run to allow Longhorn pods to start successfully
(because Longhorn pods require privileged access to manage storage).
For example, adding the following labels to the namespace that is running Longhorn pods:
    ```yaml
    apiVersion: v1
    kind: Namespace
    metadata:
      name: longhorn-system
      labels:
        pod-security.kubernetes.io/enforce: privileged
        pod-security.kubernetes.io/enforce-version: latest
        pod-security.kubernetes.io/audit: privileged
        pod-security.kubernetes.io/audit-version: latest
        pod-security.kubernetes.io/warn: privileged
   	```

### Updating CSI Snapshot CRD `v1beta1` to `v1`, `v1beta1` Deprecated

The CSI snapshot CRDs `v1beta1` version is being deprecated and replaced by `v1` version,
please follow the instruction [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and CSI snapshot controller.
If you have manifests or scripts that are still using `v1beta1` version, consider upgrading them to use `v1` as well.

```

However, Pod Security Policy has been deprecated since Kubernetes v1.21 and removed since Kubernetes v1.25, [link](https://kubernetes.io/docs/concepts/security/pod-security-policy/).
Therefore, we stopped shipping the Pod Security Policies by default.
For Kubernetes < v1.25, if your cluster still enables Pod Security Policy admission controller, set the helm value `enablePSP` to `true` to install `longhorn-psp` PodSecurityPolicy resrouce which allows privileged Longhorn pods to start.

- As a replacement for Pod Security Policy, Kubernetes provides a new mechanism, [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).
If you enable the Pod Security Admission controller and change the default behavior to block privileged pods,
you must add the correct labels to the namespace where Longhorn pods run to allow Longhorn pods to start successfully
(because Longhorn pods require privileged access to manage storage).

> For example, adding the following labels to the namespace that is running Longhorn pods:
>     ```yaml
>     apiVersion: v1
>     kind: Namespace
>     metadata:
>       name: longhorn-system
>       labels:
>         pod-security.kubernetes.io/enforce: privileged
>         pod-security.kubernetes.io/enforce-version: latest
>         pod-security.kubernetes.io/audit: privileged
>         pod-security.kubernetes.io/audit-version: latest
>         pod-security.kubernetes.io/warn: privileged
>    	```
> 
> ### Updating CSI Snapshot CRD `v1beta1` to `v1`, `v1beta1` Deprecated
> 
> The CSI snapshot CRDs `v1beta1` version is being deprecated and replaced by `v1` version,
> please follow the instruction [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and CSI snapshot controller.
> If you have manifests or scripts that are still using `v1beta1` version, consider upgrading them to use `v1` as well.
> 
>         pod-security.kubernetes.io/warn-version: latest
>     ```

Reviewer: ```suggestion
For example, adding the following labels to the namespace that is running Longhorn pods:
    ```yaml
    apiVersion: v1
    kind: Namespace
    metadata:
      name: longhorn-system
      labels:
        pod-security.kubernetes.io/enforce: privileged
        pod-security.kubernetes.io/enforce-version: latest
        pod-security.kubernetes.io/audit: privileged
        pod-security.kubernetes.io/audit-version: latest
        pod-security.kubernetes.io/warn: privileged
        pod-security.kubernetes.io/warn-version: latest
   	```

### Updating CSI Snapshot CRD `v1beta1` to `v1`, `v1beta1` Deprecated

The CSI snapshot CRDs `v1beta1` version is being deprecated and replaced by `v1` version,
please follow the instruction [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and CSI snapshot controller.
If you have manifests or scripts that are still using `v1beta1` version, consider upgrading them to use `v1` as well.
```

Please see [here](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}) for the full release note.

## Notes
1. Please ensure your Kubernetes cluster is at least v1.21 and at most v1.26 before upgrading to Longhorn v{{< current-version >}} because the supported Kubernetes version has been updated (>= v1.21 and <= v1.26) in v{{< current-version >}}.

> 1. Longhorn pods require privileged access to manage node storage. In Longhorn `v1.3.x` or older, Longhorn was shipping some Pod Security Policies by default, (i.g., [link](https://github.com/longhorn/longhorn/blob/4ba39a989b4b482d51fd4bc651f61f2b419428bd/chart/values.yaml#L260)).

Reviewer 2: what does `i.g.` stand for?

Author: It was a typo. I fixed to `e.g.,`

> However, Pod Security Policy has been deprecated since Kubernetes v1.21 and removed since Kubernetes v1.25.

Reviewer 2: Suggest linking this to https://kubernetes.io/docs/concepts/security/pod-security-policy/.

Author: Thank you. Fixed

Therefore, we stopped shipping the Pod Security Policies by default.

> For Kubernetes < v1.25, if your cluster still enables Pod Security Policy admission controller, set the helm value to `true` to ship longhorn-psp which allow privileged Longhorn pods to start.

Reviewer: ```suggestion
For Kubernetes < v1.25, if your cluster still enables Pod Security Policy admission controller, set the helm value `enablePSP` to `true` to install `longhorn-psp` PodSecurityPolicy resrouce which allows privileged Longhorn pods to start.
```

Author: Thank you. Fixed

1. As a replacement for Pod Security Policy, Kubernetes provides a new mechanism, [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/).

> If you enabled Pod Security Admission controller and change the default behavior to block privileged pods,

Reviewer: ```suggestion
If you enable the Pod Security Admission controller and change the default behavior to block privileged pods,
```

Author: Thank you. Fixed

> you must add the correct labels to the namespace that is running Longhorn pods to allow Longhorn pods to start

Reviewer: ```suggestion
you must add the correct labels to the namespace where Longhorn pods run to allow Longhorn pods to start successefully
```

Author: Thank you. Fixed

        pod-security.kubernetes.io/audit: privileged
        pod-security.kubernetes.io/audit-version: latest
        pod-security.kubernetes.io/warn: privileged
        pod-security.kubernetes.io/warn-version: latest
    ```
1. After the upgrade, the recurring job settings of volumes will be migrated to new recurring job resources, and the `RecurringJobs` field in the volume spec will be deprecated. [[doc](https://longhorn.io/docs/{{< current-version >}}/deploy/upgrade/#4-automatically-migrate-recurring-jobs)]
1. When upgrading from a Longhorn version >= 1.2.0 and <= v1.2.3 to v{{< current-version >}}, if your cluster has many backups, you may expect to have a long upgrade time (with 22000 backups, it could take a few hours). Helm upgrade may timeout and you may need to re-run the upgrade command or set a longer timeout. This is [a known issue](https://github.com/longhorn/longhorn/issues/3890)
1. The CSI snapshot CRDs `v1beta1` version is being deprecated and replaced by `v1` version,
please follow the instruction [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and CSI snapshot controller.

> If you have manifest or script that are still using `v1beta1` version, considering upgrade

Reviewer: ```suggestion
If you have manifests or scripts that are still using `v1beta1` version, consider upgrading
```

Author: Thank you. Fixed
